diff --git a/source/common/charstr.cpp b/source/common/charstr.cpp
index c4710d6e..c3562288 100644
--- a/source/common/charstr.cpp
+++ b/source/common/charstr.cpp
@@ -14,6 +14,8 @@
 *   created by: Markus W. Scherer
 */
 
+#include <cstdlib>
+
 #include "unicode/utypes.h"
 #include "unicode/putil.h"
 #include "charstr.h"
diff --git a/source/i18n/number_longnames.cpp b/source/i18n/number_longnames.cpp
index 6c7a56e5..08078c47 100644
--- a/source/i18n/number_longnames.cpp
+++ b/source/i18n/number_longnames.cpp
@@ -5,6 +5,8 @@
 
 #if !UCONFIG_NO_FORMATTING
 
+#include <cstdlib>
+
 #include "unicode/simpleformatter.h"
 #include "unicode/ures.h"
 #include "ureslocs.h"
