diff --git a/source/common/unicode/uconfig.h b/source/common/unicode/uconfig.h
index c26fce46..caf24127 100644
--- a/source/common/unicode/uconfig.h
+++ b/source/common/unicode/uconfig.h
@@ -232,6 +232,14 @@
 #   define UCONFIG_NO_REGULAR_EXPRESSIONS 1
 #endif
 
+/**
+ * \def UCONFIG_SYSTEM_COLLATION
+ * See android/system_collation/README for details.
+ */
+#ifndef UCONFIG_SYSTEM_COLLATION
+#   define UCONFIG_SYSTEM_COLLATION 0
+#endif
+
 /* common library switches -------------------------------------------------- */
 
 /**
diff --git a/source/i18n/unicode/coll.h b/source/i18n/unicode/coll.h
index 7e467df8..bf12f03a 100644
--- a/source/i18n/unicode/coll.h
+++ b/source/i18n/unicode/coll.h
@@ -54,7 +54,11 @@
 
 #include "unicode/utypes.h"
 
-#if !UCONFIG_NO_COLLATION
+#if UCONFIG_SYSTEM_COLLATION
+
+#include "../../android/system_collation/collator.h"
+
+#elif !UCONFIG_NO_COLLATION
 
 #include "unicode/uobject.h"
 #include "unicode/ucol.h"
diff --git a/source/i18n/unicode/ucol.h b/source/i18n/unicode/ucol.h
index ecabd9e0..acf9c387 100644
--- a/source/i18n/unicode/ucol.h
+++ b/source/i18n/unicode/ucol.h
@@ -12,8 +12,6 @@
 
 #include "unicode/utypes.h"
 
-#if !UCONFIG_NO_COLLATION
-
 #include "unicode/unorm.h"
 #include "unicode/localpointer.h"
 #include "unicode/parseerr.h"
@@ -21,6 +19,8 @@
 #include "unicode/uset.h"
 #include "unicode/uscript.h"
 
+#if !UCONFIG_NO_COLLATION
+
 /**
  * \file
  * \brief C API: Collator 
@@ -57,7 +57,13 @@ struct UCollator;
  */
 typedef struct UCollator UCollator;
 
+#endif /* #if !UCONFIG_NO_COLLATION */
 
+/* NOTE: The following enum declarations must be provided if:
+ * - UCONFIG_NO_COLLATION is 0
+ * - UCONFIG_NO_COLLATION is 1, and UCONFIG_SYSTEM_COLLATION is 1
+ */
+#if !UCONFIG_NO_COLLATION || UCONFIG_SYSTEM_COLLATION
 /**
  * UCOL_LESS is returned if source string is compared to be less than target
  * string in the ucol_strcoll() method.
@@ -371,6 +377,10 @@ typedef enum {
   UCOL_FULL_RULES 
 } UColRuleOption ;
 
+#endif /* #if !UCONFIG_NO_COLLATION || UCONFIG_SYSTEM_COLLATION */
+
+#if !UCONFIG_NO_COLLATION
+
 /**
  * Open a UCollator for comparing strings.
  *
