--- a/source/common/locid.cpp
+++ b/source/common/locid.cpp
@@ -662,6 +662,14 @@ Locale& Locale::init(const char* localeID, UBool canonicalize)
         }
 
         if (canonicalize) {
+            // Return for 3 common application locales which are know to be
+            // already canoncalized to speed up.
+            if (uprv_strcmp(fullName, "c") == 0 ||
+                uprv_strcmp(fullName, "en") == 0 ||
+                uprv_strcmp(fullName, "en_US") == 0) {
+                return *this;
+            }
+
             UErrorCode status = U_ZERO_ERROR;
             // TODO: Try to use ResourceDataValue and ures_getValueWithFallback() etc.
             LocalUResourceBundlePointer metadata(ures_openDirect(NULL, "metadata", &status));
