diff --git a/source/i18n/olsontz.cpp b/source/i18n/olsontz.cpp
index 00a59d16..4032b91c 100644
--- a/source/i18n/olsontz.cpp
+++ b/source/i18n/olsontz.cpp
@@ -396,7 +396,7 @@ void OlsonTimeZone::getOffset(UDate date, UBool local, int32_t& rawoff,
     if (finalZone != NULL && date >= finalStartMillis) {
         finalZone->getOffset(date, local, rawoff, dstoff, ec);
     } else {
-        getHistoricalOffset(date, local, kFormer, kLatter, rawoff, dstoff);
+        getHistoricalOffset(date, local, kFormer, kFormer, rawoff, dstoff);
     }
 }
 
diff --git a/source/i18n/rbtz.cpp b/source/i18n/rbtz.cpp
index d98613fe..5834898e 100644
--- a/source/i18n/rbtz.cpp
+++ b/source/i18n/rbtz.cpp
@@ -400,7 +400,7 @@ RuleBasedTimeZone::getOffset(uint8_t era, int32_t year, int32_t month, int32_t d
 void
 RuleBasedTimeZone::getOffset(UDate date, UBool local, int32_t& rawOffset,
                              int32_t& dstOffset, UErrorCode& status) const {
-    getOffsetInternal(date, local, kFormer, kLatter, rawOffset, dstOffset, status);
+    getOffsetInternal(date, local, kFormer, kFormer, rawOffset, dstOffset, status);
 }
 
 void
diff --git a/source/i18n/unicode/simpletz.h b/source/i18n/unicode/simpletz.h
index 5b802632..0f3835fa 100644
--- a/source/i18n/unicode/simpletz.h
+++ b/source/i18n/unicode/simpletz.h
@@ -920,7 +920,10 @@ inline void SimpleTimeZone::setEndRule(int32_t month, int32_t dayOfMonth, int32_
 inline void
 SimpleTimeZone::getOffset(UDate date, UBool local, int32_t& rawOffsetRef,
                           int32_t& dstOffsetRef, UErrorCode& ec) const {
-    TimeZone::getOffset(date, local, rawOffsetRef, dstOffsetRef, ec);
+    if (local)
+      getOffsetFromLocal(date, kFormer, kFormer, rawOffsetRef, dstOffsetRef, ec);
+    else
+      TimeZone::getOffset(date, FALSE, rawOffsetRef, dstOffsetRef, ec);
 }
 
 U_NAMESPACE_END
