diff --git a/source/common/unicode/unistr.h b/source/common/unicode/unistr.h
index 39d789fd..3f56e70c 100644
--- a/source/common/unicode/unistr.h
+++ b/source/common/unicode/unistr.h
@@ -347,7 +347,7 @@ public:
   inline bool operator==(const S &text) const {
     std::u16string_view sv(internal::toU16StringView(text));
     uint32_t len;  // unsigned to avoid a compiler warning
-    return !isBogus() && (len = length()) == sv.length() && doEquals(sv.data(), len);
+    return !isBogus() && (len = static_cast<uint32_t>(length())) == sv.length() && doEquals(sv.data(),  static_cast<int32_t>(len));
   }
 #endif  // U_HIDE_DRAFT_API
 
